cpp/src/quantiles/tdigest/tdigest_aggregation.cu

```diff
@@ -419,47 +419,51 @@ __global__ void generate_cluster_limits_kernel(int delta,
-      // during the reduction step to be trivial.
-      //
-      double adjusted_next_limit = next_limit;
-      if ((last_inserted_index < 0) ||  // if we haven't inserted anything yet
-          (nearest_w_index ==
-           last_inserted_index)) {  // if we land in the same bucket as the previous cap
-
-        // force the value into this bucket
-        nearest_w_index =
-          (last_inserted_index == group_size - 1) ? last_inserted_index : last_inserted_index + 1;
-
-        // the "adjusted" weight must be high enough so that this value will fall in the bucket.
-        // NOTE: cumulative_weight expects an absolute index into the input value stream, not a
-        // group-relative index
-        [[maybe_unused]] auto [r, i, adjusted] = cumulative_weight(nearest_w_index + group_start);
-        adjusted_next_limit                    = max(next_limit, adjusted);
-      }
-      cluster_wl[group_num_clusters[group_index]] = adjusted_next_limit;
-      last_inserted_index                         = nearest_w_index;
+    // because of the way the scale functions work, it is possible to generate clusters
+    // in such a way that we end up with "gaps" where there are no input values that
+    // fall into a given cluster.  An example would be this:
+    //
+    // cluster weight limits = 0.00003, 1.008, 3.008
+    //
+    // input values(weight) = A(1), B(2), C(3)
+    //
+    // naively inserting these values into the clusters simply by taking a lower_bound,
+    // we would get the following distribution of input values into those 3 clusters.
+    //  (), (A), (B,C)
+    //
+    // whereas what we really want is:
+    //
+    //  (A), (B), (C)
+    //
+    // to fix this, we will artificially adjust the output cluster limits to guarantee
+    // at least 1 input value will be put in each cluster during the reduction step.
+    // this does not affect final centroid results as we still use the "real" weight limits
+    // to compute subsequent clusters - the purpose is only to allow cluster selection
+    // during the reduction step to be trivial.
+    //
+    double adjusted_next_limit = next_limit;
+    int adjusted_w_index       = nearest_w_index;
+    if ((last_inserted_index < 0) ||  // if we haven't inserted anything yet
+        (nearest_w_index ==
+         last_inserted_index)) {  // if we land in the same bucket as the previous cap
+
+      // force the value into this bucket
+      adjusted_w_index = (last_inserted_index == group_size - 1)
+                           ? last_inserted_index
+                           : max(adjusted_w_index, last_inserted_index + 1);
+
+      // the "adjusted" cluster limit must be high enough so that this value will fall in the
+      // bucket. NOTE: cumulative_weight expects an absolute index into the input value stream, not
+      // a group-relative index
+      [[maybe_unused]] auto [r, i, adjusted_w] = cumulative_weight(adjusted_w_index + group_start);
```

Comment:
```suggestion
      auto adjusted_w = thrust::get<2>(cumulative_weight(adjusted_w_index + group_start));
```
Maybe extract `adjusted_w` only since `r` and `i` are not used.